toolchain/check/context.cpp

```diff
@@ -279,233 +278,375 @@ auto Context::ParamOrArgEnd(Parse::NodeKind start_kind) -> SemIR::NodeBlockId {
+        if (elements.empty()) {
+          return MakeEmptyRepresentation(node.parse_node());
+        }
 
-    case SemIR::Builtin::Kind:
-      CARBON_FATAL() << "Builtins should be named as cross-references";
+        // Find the value representation for each element, and construct a tuple
+        // of value representations.
+        llvm::SmallVector<SemIR::TypeId> value_rep_elements;
+        value_rep_elements.reserve(elements.size());
+        bool same_as_object_rep = true;
+        for (auto element_type_id : elements) {
+          // A tuple is complete if and only if all its elements are complete.
+          auto element_value_rep =
+              GetNestedValueRepresentation(element_type_id);
+          if (element_value_rep.type_id != element_type_id) {
+            same_as_object_rep = false;
+          }
+          value_rep_elements.push_back(element_value_rep.type_id);
+        }
 
-    case SemIR::PointerType::Kind:
-      return set_copy_representation(type_id);
-
-    case SemIR::ConstType::Kind: {
-      // The value representation of `const T` is the same as that of `T`.
-      // Objects are not modifiable through their value representations.
-      auto inner_value_rep = semantics_ir().GetValueRepresentation(
-          node.As<SemIR::ConstType>().inner_id);
-      if (inner_value_rep.kind == SemIR::ValueRepresentation::Unknown) {
-        return false;
+        auto value_rep = same_as_object_rep
+                             ? type_id
+                             : context.CanonicalizeTupleType(
+                                   node.parse_node(), value_rep_elements);
+        if (elements.size() == 1) {
+          // The value representation for a tuple with a single element is a
+          // tuple containing the value representation of that element.
+          // TODO: Consider doing the same for tuples with multiple small
+          // elements.
+          return MakeCopyRepresentation(value_rep);
+        }
+        // For a tuple with multiple elements, we use a pointer representation.
+        return MakePointerRepresentation(node.parse_node(), value_rep);
       }
-      semantics_ir().CompleteType(type_id, inner_value_rep);
-      return true;
+
+      case SemIR::ClassDeclaration::Kind:
+        // TODO: Support class definitions and complete class types.
+        CARBON_FATAL() << "Class types are currently never complete";
+
+      case SemIR::Builtin::Kind:
+        CARBON_FATAL() << "Builtins should be named as cross-references";
+
+      case SemIR::PointerType::Kind:
+        return MakeCopyRepresentation(type_id);
+
+      case SemIR::ConstType::Kind:
+        // The value representation of `const T` is the same as that of `T`.
+        // Objects are not modifiable through their value representations.
+        return GetNestedValueRepresentation(
+            node.As<SemIR::ConstType>().inner_id);
     }
   }
 
-  llvm_unreachable("All node kinds were handled above");
+  enum class Phase {
```

Comment:
```suggestion
  enum class Phase : int8_t {
```

https://clang.llvm.org/extra/clang-tidy/checks/performance/enum-size.html

Reply:
I'm sad about this change -- I think this makes the code communicate its intent less well, by suggesting that these enumerators have some meaning as integer values. But I suppose we need to do this in order to allow this clang-tidy check to deliver value elsewhere.

Done.